File: packages/rk_syslog/lib/src/sink.dart
/// The Dart face of the sink.
library;

import 'dart:ffi';

import 'package:ffi/ffi.dart';

import 'bindings.dart';
import 'severity.dart';
import 'status.dart';
import 'structured_data.dart';

/// A configuration, built one named key at a time.
///
/// Named keys, not a constructor with twenty optional arguments, because the
/// native side answers `unknownConfigKey` for a name it does not know — so a
/// key that this package has not heard of yet still reaches the library, and
/// a misspelled one is refused rather than ignored.
///
/// The keys, with their defaults, are listed in the README. `spool_dir` is
/// the only one that has no default.
class RkSyslogConfig {
  RkSyslogConfig({required String spoolDirectory}) {
    _values['spool_dir'] = spoolDirectory;
  }

  final Map<String, String> _values = {};

  /// Sets a key. Whether it is a key the library knows is the library's
  /// answer, given when the sink opens.
  void set(String key, String value) => _values[key] = value;

  /// Convenience for the three header fields every record carries.
  void identify({String? hostName, String? appName, String? procId}) {
    if (hostName != null) _values['host_name'] = hostName;
    if (appName != null) _values['app_name'] = appName;
    if (procId != null) _values['proc_id'] = procId;
  }

  /// The default facility for records that do not name one.
  set facility(RkFacility facility) => _values['facility'] = facility.wireName;

  /// An unmodifiable view, for tests and for logging what was asked for.
  Map<String, String> get entries => Map.unmodifiable(_values);
}

/// A running syslog sink.
///
/// # Which isolate
///
/// И145: no call into a native library runs on the interface isolate.
/// [submit] is cheap and never waits on a network, but it is still a foreign
/// call and belongs on a worker isolate with everything else in the layer
/// below the contract. [flush] blocks on purpose and must never be anywhere
/// near the interface.
///
/// # What it guarantees
///
/// Records leave in the order they were submitted, across a restart, and are
/// delivered **at least once** — see the crate documentation for why syslog
/// cannot promise exactly once.
class RkSyslogSink {
  RkSyslogSink._(this._bindings, this._handle);

  final RkSyslogBindings _bindings;
  Pointer<RkSinkHandle> _handle;
  bool _closed = false;

  /// Whether the native library can be loaded in this build.
  ///
  /// Answers rather than throws: a web build, or a build whose native library
  /// has not been wired up yet, gets `false` and can say so in its own words.
  static bool isAvailable({String? libraryPath}) =>
      openRkSyslogLibrary(explicitPath: libraryPath) != null;

  /// The version of the loaded native library, or `null` if there is none.
  static String? nativeVersion({String? libraryPath}) {
    final library = openRkSyslogLibrary(explicitPath: libraryPath);
    if (library == null) return null;
    return RkSyslogBindings(library).version().toDartString();
  }

  /// Opens a sink.
  ///
  /// Everything the native side can check up front — the spool directory, the
  /// bound, the certificates, every configuration value — is checked here, so
  /// a mistake surfaces once at startup rather than on every record.
  static RkSyslogResult<RkSyslogSink> open(
    RkSyslogConfig config, {
    String? libraryPath,
  }) {
    final library = openRkSyslogLibrary(explicitPath: libraryPath);
    if (library == null) {
      return RkSyslogFailure(
        RkSyslogStatus.unrecognised,
        'the rk_syslog native library could not be loaded; looked for '
        '${rkSyslogLibraryFileName()}, the RK_SYSLOG_LIB environment '
        'variable, and symbols already in this process',
      );
    }
    final RkSyslogBindings bindings;
    try {
      bindings = RkSyslogBindings(library);
    } on ArgumentError catch (error) {
      return RkSyslogFailure(
        RkSyslogStatus.unrecognised,
        'the loaded library is missing a symbol this package needs, so it is '
        'not the rk_syslog this binding was built against: $error',
      );
    }

    final handle = bindings.configNew();
    if (handle == nullptr) {
      return const RkSyslogFailure(
        RkSyslogStatus.panicked,
        'the native library could not allocate a configuration',
      );
    }
    try {
      for (final entry in config.entries.entries) {
        final outcome = _call(
          bindings,
          (error) => _withStrings(
            [entry.key, entry.value],
            (pointers) =>
                bindings.configSet(handle, pointers[0], pointers[1], error),
          ),
        );
        if (outcome case final RkSyslogFailure<void> failure) {
          return failure.cast<RkSyslogSink>();
        }
      }

      final out = calloc<Pointer<RkSinkHandle>>();
      try {
        final outcome = _call(
          bindings,
          (error) => bindings.open(handle, out, error),
        );
        if (outcome case final RkSyslogFailure<void> failure) {
          return failure.cast<RkSyslogSink>();
        }
        return RkSyslogOk(RkSyslogSink._(bindings, out.value));
      } finally {
        calloc.free(out);
      }
    } finally {
      // Ours to free: we made it (И146). The native side only read it.
      bindings.configFree(handle);
    }
  }

  /// Submits one record.
  ///
  /// Frames it and hands it to the worker. Does not touch the disk or the
  /// network on this thread — a collector that is down cannot slow this down.
  ///
  /// A record that cannot be framed comes back as a failure, here, now, with
  /// the offending field named. It is never quietly dropped.
  ///
  /// [timestamp] defaults to now, with this machine's UTC offset.
  RkSyslogResult<void> submit({
    required RkSeverity severity,
    String? message,
    RkFacility? facility,
    DateTime? timestamp,
    String? msgid,
    RkStructuredData? structuredData,
  }) {
    if (_closed) {
      return const RkSyslogFailure(
        RkSyslogStatus.closed,
        'this sink has been closed',
      );
    }
    final when = timestamp ?? DateTime.now();
    final micros = when.toUtc().microsecondsSinceEpoch;
    final offsetMinutes = when.isUtc ? 0 : when.timeZoneOffset.inMinutes;

    Pointer<RkSdHandle> sd = nullptr;
    try {
      if (structuredData != null && structuredData.isNotEmpty) {
        sd = _bindings.sdNew();
        if (sd == nullptr) {
          return const RkSyslogFailure(
            RkSyslogStatus.panicked,
            'the native library could not allocate structured data',
          );
        }
        final built = structuredData.writeInto(_bindings, sd);
        if (built case final RkSyslogFailure<void> failure) {
          return failure;
        }
      }
      return _call(
        _bindings,
        (error) => _withStrings(
          [severity.wireName, facility?.wireName, msgid, message],
          (pointers) => _bindings.submit(
            _handle,
            pointers[0],
            pointers[1],
            micros,
            offsetMinutes,
            pointers[2],
            sd,
            pointers[3],
            error,
          ),
        ),
      );
    } finally {
      if (sd != nullptr) _bindings.sdFree(sd);
    }
  }

  /// Waits until everything submitted so far has reached the spool.
  ///
  /// **Blocks.** The only call here that does, and only the isolate that asks.
  /// Use it at shutdown, and wherever a record must be on disk before the next
  /// thing happens. It says nothing about delivery: a collector may be
  /// unreachable for a week and this still returns.
  RkSyslogResult<void> flush({Duration timeout = const Duration(seconds: 5)}) {
    if (_closed) {
      return const RkSyslogFailure(
        RkSyslogStatus.closed,
        'this sink has been closed',
      );
    }
    return _call(
      _bindings,
      (error) => _bindings.flush(_handle, timeout.inMilliseconds, error),
    );
  }

  /// Reads a counter by name.
  ///
  /// An unknown name comes back as [RkSyslogStatus.unknownStat], not as zero:
  /// "not measured" and "measured, and none" are different answers and a
  /// dashboard that confuses them is worse than one with a gap in it.
  ///
  /// The names are listed in the README; [counterNames] asks the library.
  RkSyslogResult<int> stat(String name) {
    if (_closed) {
      return const RkSyslogFailure(
        RkSyslogStatus.closed,
        'this sink has been closed',
      );
    }
    final out = calloc<Int64>();
    try {
      final outcome = _call(
        _bindings,
        (error) => _withStrings([
          name,
        ], (pointers) => _bindings.stat(_handle, pointers[0], out, error)),
      );
      return switch (outcome) {
        RkSyslogOk<void>() => RkSyslogOk(out.value),
        final RkSyslogFailure<void> failure => failure.cast<int>(),
      };
    } finally {
      calloc.free(out);
    }
  }

  /// Every counter name the loaded library publishes.
  List<String> get counterNames => _names(_bindings, 'counter');

  /// Every configuration key the loaded library accepts.
  List<String> get configKeys => _names(_bindings, 'config_key');

  /// Closes the sink: stops accepting, waits for the worker to finish writing
  /// what it accepted, then frees.
  ///
  /// Deterministic, and it must be called (И146). Dart's collector does not
  /// know about the spool, the worker thread or the socket, and will not
  /// close them for you. Calling it twice is safe.
  void close() {
    if (_closed) return;
    _closed = true;
    _bindings.close(_handle);
    _handle = nullptr;
  }

  /// Checks this package's name tables against the loaded library's.
  ///
  /// The reason enums cross by name (И147) is that the two sides can be
  /// compared. This is the comparison. It returns the disagreements; an empty
  /// list means the tables match.
  static List<String> verifyNameTables({String? libraryPath}) {
    final library = openRkSyslogLibrary(explicitPath: libraryPath);
    if (library == null) return const ['the native library is not loaded'];
    final bindings = RkSyslogBindings(library);
    final problems = <String>[];

    void compare(String kind, List<String> ours) {
      final theirs = _names(bindings, kind);
      for (final name in ours) {
        if (!theirs.contains(name)) {
          problems.add(
            "this package knows '$name' as a $kind; the library "
            'does not',
          );
        }
      }
      for (final name in theirs) {
        if (!ours.contains(name)) {
          problems.add(
            "the library knows '$name' as a $kind; this package "
            'does not',
          );
        }
      }
    }

    compare('severity', [for (final s in RkSeverity.values) s.wireName]);
    compare('facility', [for (final f in RkFacility.values) f.wireName]);
    compare('status', [
      for (final s in RkSyslogStatus.values)
        if (s != RkSyslogStatus.unrecognised) s.wireName,
    ]);

    // Names agreeing is not enough: the numbers behind them must agree too,
    // or a record would carry the right word and the wrong PRI.
    final out = calloc<Int32>();
    try {
      for (final severity in RkSeverity.values) {
        final name = severity.wireName.toNativeUtf8();
        try {
          if (bindings.severityValue(name, out) != 0 ||
              out.value != severity.code) {
            problems.add(
              "severity '${severity.wireName}' is ${severity.code} "
              'here and ${out.value} in the library',
            );
          }
        } finally {
          calloc.free(name);
        }
      }
      for (final facility in RkFacility.values) {
        final name = facility.wireName.toNativeUtf8();
        try {
          if (bindings.facilityValue(name, out) != 0 ||
              out.value != facility.code) {
            problems.add(
              "facility '${facility.wireName}' is ${facility.code} "
              'here and ${out.value} in the library',
            );
          }
        } finally {
          calloc.free(name);
        }
      }
    } finally {
      calloc.free(out);
    }
    return problems;
  }

  /// Provokes a panic behind the boundary, to prove for yourself that one
  /// comes back as [RkSyslogStatus.panicked] rather than taking the process
  /// with it (И144).
  static RkSyslogResult<void> provokePanicForTesting({String? libraryPath}) {
    final library = openRkSyslogLibrary(explicitPath: libraryPath);
    if (library == null) {
      return const RkSyslogFailure(
        RkSyslogStatus.unrecognised,
        'the native library is not loaded',
      );
    }
    final bindings = RkSyslogBindings(library);
    return _call(bindings, bindings.provokePanic);
  }
}

// --------------------------------------------------------------- plumbing

List<String> _names(RkSyslogBindings bindings, String kind) {
  final kindPointer = kind.toNativeUtf8();
  try {
    final out = <String>[];
    for (var index = 0; ; index++) {
      final pointer = bindings.nameAt(kindPointer, index);
      if (pointer == nullptr) break;
      out.add(pointer.toDartString());
      if (index > 4096) break; // a library that never ends is a broken one
    }
    return out;
  } finally {
    calloc.free(kindPointer);
  }
}

/// Runs a native call that reports through an out-parameter, and turns its
/// code and detail into a result.
///
/// The detail string is allocated by the native side and freed by the native
/// side (И146) — this function is the only place that happens, so there is
/// one rule and one place to check it.
RkSyslogResult<void> _call(
  RkSyslogBindings bindings,
  int Function(Pointer<Pointer<Utf8>>) body,
) {
  final error = calloc<Pointer<Utf8>>();
  try {
    final code = body(error);
    if (code == 0) return const RkSyslogOk(null);
    final namePointer = bindings.statusName(code);
    final status = RkSyslogStatus.fromName(
      namePointer == nullptr ? null : namePointer.toDartString(),
    );
    var detail = '';
    if (error.value != nullptr) {
      detail = error.value.toDartString();
      bindings.stringFree(error.value);
      error.value = nullptr;
    }
    if (status == RkSyslogStatus.unrecognised) {
      detail = detail.isEmpty
          ? 'the library returned status $code, which this package has no '
                'name for; it is newer than this binding'
          : '$detail (library status $code, unknown to this binding)';
    }
    return RkSyslogFailure(status, detail);
  } finally {
    calloc.free(error);
  }
}

/// Converts strings to native memory for the duration of one call, then frees
/// them. A null entry becomes `nullptr`, which every entry point reads as
/// "absent".
///
/// The native side copies whatever it keeps, so freeing here is correct and
/// necessary: nothing on the other side outlives the call.
int _withStrings(List<String?> values, int Function(List<Pointer<Utf8>>) body) {
  final pointers = <Pointer<Utf8>>[];
  try {
    for (final value in values) {
      pointers.add(value == null ? nullptr : value.toNativeUtf8());
    }
    return body(pointers);
  } finally {
    for (final pointer in pointers) {
      if (pointer != nullptr) calloc.free(pointer);
    }
  }
}

File: packages/rk_syslog/rust/src/sink.rs
//! The sink: framing on the caller's thread, everything else on a worker.
//!
//! # What is guaranteed, and what is not
//!
//! **Order.** Records leave in the order they were submitted, across a crash
//! and a restart. The spool is a file read front to back and a cursor that
//! only moves forward.
//!
//! **At-least-once, not exactly-once.** The cursor advances after a record
//! has been written to the socket and flushed. A machine that dies in the
//! gap between the write and the cursor reaching disk resends that record on
//! restart, and the collector sees it twice. The alternative — advancing
//! first — turns the same gap into a lost record, and for a journal a
//! duplicate is a nuisance while a hole is a defect. There is nothing in
//! RFC 5425 to make this exact: the protocol has no application-level
//! acknowledgement to make "delivered" a thing either end can agree on.
//!
//! **Durability begins at the spool, not at submit.** See [`crate::queue`].
//!
//! **Gaps are counted, never silent.** Under the drop-oldest rule the worker
//! writes a record of its own into the spool saying how many were discarded,
//! so a reader of the journal sees the hole in the journal rather than having
//! to think to ask a counter.

use std::sync::atomic::{AtomicU64, Ordering};
use std::sync::Arc;
use std::thread::JoinHandle;
use std::time::{Duration, Instant, SystemTime, UNIX_EPOCH};

use crate::config::Settings;
use crate::queue::Queue;
use crate::rfc5424::{frame, Record, StructuredData, RK_SD_ID};
use crate::severity::Severity;
use crate::spool::Spool;
use crate::status::{Failure, Fallible, Status};
use crate::transport::Transport;

/// Counters, looked up by name.
///
/// By name for the same reason the enums are: a binding that reads counter
/// three gets a different number every time the list changes, and nothing
/// tells it. Asking for `"spool_dropped_records"` either answers or says it
/// does not know that name.
#[derive(Debug, Default)]
pub struct Counters {
    submitted: AtomicU64,
    framing_refused: AtomicU64,
    truncated: AtomicU64,
    queue_refused: AtomicU64,
    queue_dropped: AtomicU64,
    queue_depth: AtomicU64,
    spooled: AtomicU64,
    spool_refused: AtomicU64,
    spool_dropped_records: AtomicU64,
    spool_dropped_segments: AtomicU64,
    spool_torn_records: AtomicU64,
    spool_bytes: AtomicU64,
    spool_records: AtomicU64,
    spool_io_failures: AtomicU64,
    sent: AtomicU64,
    send_failures: AtomicU64,
    connected: AtomicU64,
}

/// Every counter name this version publishes.
pub const COUNTER_NAMES: &[&str] = &[
    "submitted",
    "framing_refused",
    "truncated",
    "queue_refused",
    "queue_dropped",
    "queue_depth",
    "spooled",
    "spool_refused",
    "spool_dropped_records",
    "spool_dropped_segments",
    "spool_torn_records",
    "spool_bytes",
    "spool_records",
    "spool_io_failures",
    "sent",
    "send_failures",
    "connected",
];

impl Counters {
    fn cell(&self, name: &str) -> Option<&AtomicU64> {
        Some(match name {
            "submitted" => &self.submitted,
            "framing_refused" => &self.framing_refused,
            "truncated" => &self.truncated,
            "queue_refused" => &self.queue_refused,
            "queue_dropped" => &self.queue_dropped,
            "queue_depth" => &self.queue_depth,
            "spooled" => &self.spooled,
            "spool_refused" => &self.spool_refused,
            "spool_dropped_records" => &self.spool_dropped_records,
            "spool_dropped_segments" => &self.spool_dropped_segments,
            "spool_torn_records" => &self.spool_torn_records,
            "spool_bytes" => &self.spool_bytes,
            "spool_records" => &self.spool_records,
            "spool_io_failures" => &self.spool_io_failures,
            "sent" => &self.sent,
            "send_failures" => &self.send_failures,
            "connected" => &self.connected,
            _ => return None,
        })
    }

    pub fn get(&self, name: &str) -> Option<u64> {
        self.cell(name).map(|c| c.load(Ordering::Relaxed))
    }

    fn bump(&self, name: &str, by: u64) {
        if let Some(cell) = self.cell(name) {
            cell.fetch_add(by, Ordering::Relaxed);
        }
    }

    fn set(&self, name: &str, value: u64) {
        if let Some(cell) = self.cell(name) {
            cell.store(value, Ordering::Relaxed);
        }
    }
}

/// A running syslog sink.
#[derive(Debug)]
pub struct Sink {
    settings: Settings,
    queue: Arc<Queue>,
    counters: Arc<Counters>,
    worker: Option<JoinHandle<()>>,
}

impl Sink {
    /// Opens the spool, checks the collector settings and starts the worker.
    ///
    /// Everything that can fail for a reason the caller can fix — an
    /// unwritable directory, a certificate that is not there, a bound smaller
    /// than a segment — fails here, as a returned value, before a single
    /// record has been submitted.
    pub fn open(settings: Settings) -> Fallible<Sink> {
        let spool = Spool::open(std::path::Path::new(&settings.spool_dir), settings.spool)?;
        let transport = Transport::new(settings.collector.clone())?;

        let counters = Arc::new(Counters::default());
        let stats = spool.stats();
        counters.set("spool_bytes", stats.bytes);
        counters.set("spool_records", stats.records);
        counters.set("spool_torn_records", stats.torn_records);

        let queue = Arc::new(Queue::new(settings.queue_capacity, settings.queue_policy));

        let worker = {
            let queue = Arc::clone(&queue);
            let counters = Arc::clone(&counters);
            let settings = settings.clone();
            std::thread::Builder::new()
                .name("rk_syslog".into())
                .spawn(move || run_worker(settings, queue, counters, spool, transport))
                .map_err(|e| {
                    Failure::new(
                        Status::SpoolIo,
                        format!("starting the rk_syslog worker thread: {e}"),
                    )
                })?
        };

        Ok(Sink {
            settings,
            queue,
            counters,
            worker: Some(worker),
        })
    }

    /// Frames a record and hands it to the worker.
    ///
    /// Touches no socket and no file. The only failures it can return are the
    /// ones it can decide here and now: a record that cannot be framed, and a
    /// hand-off queue that is full.
    pub fn submit(&self, record: &Record) -> Fallible<()> {
        self.counters.bump("submitted", 1);
        let framed = match frame(
            &self.settings.identity,
            record,
            self.settings.max_message_bytes,
            self.settings.oversize,
            self.settings.emit_bom,
        ) {
            Ok(framed) => framed,
            Err(failure) => {
                self.counters.bump("framing_refused", 1);
                return Err(failure);
            }
        };
        if framed.truncated_from.is_some() {
            self.counters.bump("truncated", 1);
        }
        match self.queue.push(framed.bytes) {
            Ok(()) => {
                self.counters.set("queue_depth", self.queue.len() as u64);
                Ok(())
            }
            Err(failure) => {
                if failure.status == Status::QueueFull {
                    self.counters.bump("queue_refused", 1);
                }
                Err(failure)
            }
        }
    }

    /// Waits until everything submitted so far has reached the spool.
    ///
    /// The one call here that blocks, and only the caller who asks for it.
    /// Use it at shutdown, and at any point where a record must be on disk
    /// before the next thing happens. It says nothing about delivery — the
    /// collector may be unreachable for days and this still returns.
    pub fn flush(&self, timeout: Duration) -> Fallible<()> {
        self.queue.wait_drained(timeout)
    }

    pub fn stat(&self, name: &str) -> Option<u64> {
        if name == "queue_depth" {
            return Some(self.queue.len() as u64);
        }
        self.counters.get(name)
    }

    pub fn settings(&self) -> &Settings {
        &self.settings
    }
}

impl Drop for Sink {
    /// Closes the queue and joins the worker, so a closed sink has written
    /// everything it accepted. Deterministic on purpose: nothing here waits
    /// for a garbage collector to notice.
    fn drop(&mut self) {
        self.queue.close();
        if let Some(worker) = self.worker.take() {
            let _ = worker.join();
        }
    }
}

/// How long the worker sleeps when there is nothing to do. Short enough that
/// a retry after a failed send is not delayed by an idle queue.
const IDLE_WAIT: Duration = Duration::from_millis(250);

/// Records sent per pass before going back to look at the queue, so a large
/// backlog cannot starve the hand-off.
const SEND_BATCH: usize = 256;

fn run_worker(
    settings: Settings,
    queue: Arc<Queue>,
    counters: Arc<Counters>,
    mut spool: Spool,
    mut transport: Transport,
) {
    let mut backoff = settings.retry_min;
    let mut next_attempt = Instant::now();

    loop {
        let (batch, closed) = queue.drain_blocking(IDLE_WAIT);
        for record in &batch {
            match spool.append(record) {
                Ok(()) => counters.bump("spooled", 1),
                Err(failure) => {
                    if failure.status == Status::SpoolFull {
                        counters.bump("spool_refused", 1);
                    } else {
                        counters.bump("spool_io_failures", 1);
                    }
                }
            }
        }
        if !batch.is_empty() {
            let _ = spool.flush();
        }
        queue.mark_drained();

        let queue_dropped = queue.take_dropped();
        if queue_dropped > 0 {
            counters.bump("queue_dropped", queue_dropped);
            announce_loss(&settings, &mut spool, "queueDropped", queue_dropped);
        }

        let dropped = spool.take_dropped();
        if dropped > 0 {
            counters.bump("spool_dropped_records", dropped);
            // The hole is written into the journal, not only into a counter.
            announce_loss(&settings, &mut spool, "spoolDropped", dropped);
        }

        let stats = spool.stats();
        counters.set("spool_bytes", stats.bytes);
        counters.set("spool_records", stats.records);
        counters.set("spool_dropped_segments", stats.dropped_segments);
        counters.set("spool_torn_records", stats.torn_records);
        counters.set("queue_depth", queue.len() as u64);

        if transport.is_enabled() && Instant::now() >= next_attempt {
            let mut sent = 0usize;
            loop {
                if sent >= SEND_BATCH {
                    break;
                }
                let record = match spool.peek() {
                    Ok(Some(record)) => record,
                    Ok(None) => break,
                    Err(_) => break,
                };
                match transport.send(&record) {
                    Ok(()) => {
                        // Only now does the cursor move. A crash before this
                        // line resends the record; a crash after it does not.
                        // See the at-least-once note at the top of this file.
                        if spool.advance(record.len()).is_err() {
                            break;
                        }
                        counters.bump("sent", 1);
                        counters.set("connected", 1);
                        sent += 1;
                        backoff = settings.retry_min;
                    }
                    Err(_) => {
                        counters.bump("send_failures", 1);
                        counters.set("connected", 0);
                        transport.disconnect();
                        next_attempt = Instant::now() + backoff;
                        backoff = (backoff * 2).min(settings.retry_max);
                        break;
                    }
                }
            }
        }

        if closed && queue.is_empty() {
            let _ = spool.flush();
            return;
        }
    }
}

/// Writes a record saying that records were lost, into the place the lost
/// records would have been.
fn announce_loss(settings: &Settings, spool: &mut Spool, what: &str, count: u64) {
    let micros = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_micros() as i64)
        .unwrap_or(0);
    let mut sd = StructuredData::new();
    if sd.push_element(RK_SD_ID).is_err() {
        return;
    }
    if sd.push_param(what, &count.to_string()).is_err() {
        return;
    }
    let record = Record {
        facility: settings.facility,
        severity: Severity::Warning,
        epoch_micros: micros,
        utc_offset_minutes: 0,
        msgid: Some("RKLOSS".into()),
        structured_data: sd,
        message: Some(format!(
            "rk_syslog discarded {count} record(s) to stay inside its bound ({what})"
        )),
    };
    if let Ok(framed) = frame(
        &settings.identity,
        &record,
        settings.max_message_bytes,
        settings.oversize,
        settings.emit_bom,
    ) {
        let _ = spool.append(&framed.bytes);
    }
}
